src/org/labkey/filetransfer/FileTransferController.java

```diff
@@ -119,10 +151,362 @@ public URLHelper getSuccessURL(FileTransferConfigForm form)
+            this.dataRegionSelectionKey = dataRegionSelectionKey;
+        }
+
+        public Integer getWebPartId()
+        {
+            return webPartId;
+        }
+
+        public void setWebPartId(Integer webPartId)
+        {
+            this.webPartId = webPartId;
+        }
+    }
+
+    /**
+     * This action is the target of the authorization action from the file transfer provider.  If authorization has
+     * been granted, the code provided from that authorization will be used to retrieve credentials to be used in
+     * initiating transfer requests.  In any case, this action redirects to a page where we give feedback to the user
+     * and display items for next steps, which are either to return to the page where the initial selection of files was
+     * made, choose a destination endpoint, or initiate a transfer of the selected files.
+     */
+    @CSRF
+    @RequiresPermission(ReadPermission.class)
+    public class TokensAction extends RedirectAction<AuthForm>
+    {
+        Boolean authorized = true;
+        FileTransferManager.ErrorCode errorCode = null;
+
+        @Override
+        public URLHelper getSuccessURL(AuthForm authForm)
+        {
+            ActionURL url = new ActionURL(PrepareAction.class, getContainer()).addParameter("authorized", authorized);
+            if (errorCode != null)
+                url.addParameter("errorCode", errorCode.toString());
+            String returnUrl = (String) getViewContext().getSession().getAttribute(FileTransferManager.RETURN_URL_SESSION_KEY);
+            if (returnUrl != null)
+                try
+                {
+                    url.addReturnURL(new URLHelper(returnUrl));
+                }
+                catch (URISyntaxException e)
+                {
+                    logger.error("Invalid URI syntax for returnUrl " + returnUrl + " returning to container start URL", e);
+                    url.addReturnURL(getViewContext().getContainer().getStartURL(getUser()));
+                }
+            return url;
+        }
+
+        @Override
+        public boolean doAction(AuthForm form, BindException errors) throws Exception
+        {
+            if (form.getError() != null)
+            {
+                // not an error because the user may have simply chosen not to authorize the client.
+                logger.info("Error in authorizing: " + form.getError());
+                this.authorized = false;
+                return true;
+            }
+            else if (form.getCode() != null)
+            {
+                SecurePropertiesDataStore store = new SecurePropertiesDataStore(getUser(), getContainer());
+
+                // TODO check if this actually retrieves from the database or if there's more going on
+//                StoredCredential credential = store.get(null);
+//                if (credential.getAccessToken() == null || credential.getExpirationTimeMilliseconds() == null || credential.getExpirationTimeMilliseconds() <= 0)
```

Comment:
Should this commented out code be removed, or is it something to be handled in a future sprint?